### PR TITLE
#159965267 Implement social login strategies

### PR DESCRIPTION
**Pull request by pngg**
#### What does this PR do?
Implementation of social login feature for Authors Haven

#### Description of Task to be completed?
The social login feature allows a user to log in via their favourite social platforms.

#### How should this be manually tested?
Clone: `https://github.com/andela/mag-ah-frontend.git`
Checkout to: `ft-social-auth-159965267`
Install requirements: `npm install`
Run app: `npm start`

Navigate to the following url:

`http://localhost:3000/` and click on `Get Started` link.

Click on your favourite social platform button
screenshot 2018-10-19 14 35 53

#### Any background context you want to provide?
Twitter login strategy is still a WIP because we are still troubleshooting the implementation.

#### What are the relevant pivotal tracker stories?
[Users should be able to Login via Google Facebook or Twitter](https://www.pivotaltracker.com/story/show/159965267)

### CODE REVIEW

**Comment by kimotho-njoki** [comment]
@pngg I can see Travis is failing and conflicting files, when would you like this reviewed?

**Comment by pngg** [comment]
> @pngg I can see Travis is failing and conflicting files, when would you like this reviewed?

@kimotho-njoki Trying to fix this and update the PR in the next hour.

File: src/__tests__/jest/actions/socialLoginAction.test.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import configureMockStore from "redux-mock-store";
+
+import { socialLoginSuccess } from "../../../redux/actions/SocialLoginAction";
+import * as actionTypes from "../../../redux/action_types";
+import { socialPayload } from "../../mock/mockData";
+
+const middleware = [thunk];
+const mockStore = configureMockStore(middleware);
+const loginURL =
+  "https://ah-magnificent6-staging.herokuapp.com/api/users/oauth/";
```
Comment (Bileonaire) [inline src/__tests__/jest/actions/socialLoginAction.test.js]: Use the config file and .env to set up the BASE_URL

**Review by Bileonaire — CHANGES_REQUESTED** [review]
The PR Looks good to me

File: src/__tests__/jest/actions/socialLoginAction.test.js (hunk 1 above)
Comment (pngg) [reply src/__tests__/jest/actions/socialLoginAction.test.js]: @Bileonaire This has been fixed.

**Review by Bileonaire — APPROVED** [review]

**Comment by micahoriaso** [comment]
This looks good to me

**Comment by micahoriaso** [comment]
This looks good to me

**Review by micahoriaso — COMMENTED** [review]
This looks good to me

**Review by micahoriaso — COMMENTED** [review]
This looks good to me

**Comment by micahoriaso** [comment]
This looks good to me

**Comment by micahoriaso** [comment]
This looks good to me

**Comment by micahoriaso** [comment]
Looks good to me

File: src/__tests__/jest/actions/socialLoginAction.test.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+const middleware = [thunk];
+const mockStore = configureMockStore(middleware);
+
+const loginURL = config.BASE_URL;
+
+describe("socialLoginAction", () => {
+  beforeEach(() => moxios.install());
+  afterEach(() => moxios.uninstall());
+
+  it("should dispatch successful social login", async done => {
```
Comment (kimotho-njoki) [inline src/__tests__/jest/actions/socialLoginAction.test.js]: `done` here is acting as a placeholder for () as seen on line 44. Please delete the line 44 and replace `done` with `()` on line 19. No need for unnecessary extra lines of code. Do the same for all files you have included done.

Comment (kimotho-njoki) [inline src/__tests__/jest/actions/socialLoginAction.test.js]: `done` here is acting as a placeholder for () as seen on line 44. Please delete the line 44 and replace `done` with `()` on line 19. No need for unnecessary extra lines of code. Do the same for all files you have included done.

File: src/views/Login/AuthButton.jsx [hunk 3]
```diff
@@ -0,0 +1,50 @@
+import React from "react";
+import PropTypes from "prop-types";
+import { connect } from "react-redux";
+import Button from "../Button";
+import { Logout } from "../../redux/actions/loginAction";
+
+class AuthButton extends React.Component {
+  constructor(props) {
+    super(props);
+    this.onLogout = this.onLogout.bind(this);
```
Comment (kimotho-njoki) [inline src/views/Login/AuthButton.jsx]: If you change the `onLogout()` to an ES6 function you will not need the bind statement nor the constructor. Read more here https://medium.freecodecamp.org/react-binding-patterns-5-approaches-for-handling-this-92c651b5af56?gi=8207f4fed723

Comment (pngg) [reply src/views/Login/AuthButton.jsx]: @kimotho-njoki Noted

File: src/views/Login/login.jsx [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+class LoginPage extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = { email: "", password: "" };
+    this.handleChange = this.handleChange.bind(this);
+    this.onLogin = this.onLogin.bind(this);
+  }
+
+  onLogin(event) {
```
Comment (kimotho-njoki) [inline src/views/Login/login.jsx]: Use the same link and feedback I gave you up there to refactor this as well, remove all bind statements.

File: src/views/Login/login.jsx [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    );
+  }
+}
+
+LoginPage.propTypes = {
+  dispatch: PropTypes.func.isRequired,
+  error: PropTypes.shape().isRequired
+};
+
+const mapStateToProps = ({ loginReducer }) => loginReducer;
```
Comment (kimotho-njoki) [inline src/views/Login/login.jsx]: What are you getting from the loginReducer? Format in such a way that we all understand what you are trying to access.

const mapStateToProps = ({ loginReducer }) => {
const { whatYouNeed } = loginReducer;
return {
whatYouNeed
};
};

File: src/views/Login/socialAuth.jsx [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+      </div>
+    );
+  }
+}
+
+SocialLogin.propTypes = {
+  dispatch: PropTypes.func.isRequired
+};
+
+const mapStateToProps = ({ loginReducer }) => loginReducer;
```
Comment (kimotho-njoki) [inline src/views/Login/socialAuth.jsx]: Refactor this mapStateToProps as well

**Comment by kimotho-njoki** [comment]
@pngg I've tested this and `Twitter` is the only option that seems to redirect me, `Facebook` and `Google` links do not seem to be working. I thought from what you told me on PT you had implemented both `Google` and `Facebook` but unable to implement `Twitter`.

**Comment by pngg** [comment]
> @pngg I've tested this and `Twitter` is the only option that seems to redirect me, `Facebook` and `Google` links do not seem to be working. I thought from what you told me on PT you had implemented both `Google` and `Facebook` but unable to implement `Twitter`.

@kimotho-njoki Yes both Google and Facebook are working but they are reading `keys` from .env file which is not available on github. I have however included a .env-sample file for configurations.

File: src/views/Login/login.jsx [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+import { connect } from "react-redux";
+import loginAction from "../../redux/actions/loginAction";
+import Button from "../Button";
+import LoginForm from "./LoginForm";
+import SocialAuth from "./socialAuth";
+
+class LoginPage extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = { email: "", password: "" };
```
Comment (kimotho-njoki) [inline src/views/Login/login.jsx]: The constructor is now irrelevant as the bind statement has been removed. Refactor to remain only with the state.

Comment (pngg) [reply src/views/Login/login.jsx]: @kimotho-njoki I have removed the constructor method.

File: src/views/Login/login.jsx [hunk 8]
```diff
@@ -0,0 +1,68 @@
+import React from "react";
+import PropTypes from "prop-types";
+import { connect } from "react-redux";
+import loginAction from "../../redux/actions/loginAction";
+import Button from "../Button";
+import LoginForm from "./LoginForm";
+import SocialAuth from "./socialAuth";
+
+class LoginPage extends React.Component {
+  state = { email: "", password: "" };
```
Comment (kimotho-njoki) [inline src/views/Login/login.jsx]: Refactor to make it more readable
state = {
email: "",
password: ""
}

**Comment by kimotho-njoki** [comment]
one more nitpick

**Pull request merged**